awx/main/fields.py

```diff
@@ -858,27 +856,41 @@ def __str__(self):
                 template_name = template_name.split('.')[1]
                 setattr(valid_namespace['tower'].filename, template_name, 'EXAMPLE_FILENAME')
 
+        def validate_template_string(type_, key, tmpl):
+            try:
+                sandbox.ImmutableSandboxedEnvironment(undefined=StrictUndefined).from_string(tmpl).render(valid_namespace)
+            except UndefinedError as e:
+                raise django_exceptions.ValidationError(
+                    _('{sub_key} uses an undefined field ({error_msg})').format(sub_key=key, error_msg=e),
+                    code='invalid',
+                    params={'value': value},
+                )
+            except SecurityError as e:
+                raise django_exceptions.ValidationError(_('Encountered unsafe code execution: {}').format(e))
+            except TemplateSyntaxError as e:
+                raise django_exceptions.ValidationError(
+                    _('Syntax error rendering template for {sub_key} inside of {type} ({error_msg})').format(sub_key=key, type=type_, error_msg=e),
+                    code='invalid',
+                    params={'value': value},
+                )
+
+        def validate_extra_vars(key, node):
+            if isinstance(node, dict):
+                return {validate_extra_vars(key, k): validate_extra_vars(k, v) for k, v in node.items()}
+            elif isinstance(node, list):
+                return [validate_extra_vars(key, x) for x in node]
```

Comment:
I'd like to change `for x in node` to `for i, x in enumerate(node)` and then stack onto `key`, like `f'{key}:{i}`. It doesn't distinguish between dicts and lists... but it shouldn't matter because the user know what injectors they submitted.

Reply:
I think the context is fully maintained now in the validation error messages. I used `key[i][i]` style for the validation error message, as between the dict dot style and this, it should be pseudo jq/jmespath style.

@AlanCoding: running `podman exec -it awx-tools_awx_1 make test_unit` works passes the new unit tests, But I'm not sure how to actually test those validation error paths.